p4_symbolic/ir/ir.cc

```diff
@@ -0,0 +1,354 @@
+  std::vector<std::string> variable_map(bmv2_action.runtime_data_size());
+  for (int i = 0; i < bmv2_action.runtime_data_size(); i++) {
+    const bmv2::VariableDefinition variable = bmv2_action.runtime_data(i);
+    (*action_impl->mutable_variables())[variable.name()] = variable.bitwidth();
+    variable_map[i] = variable.name();
+  }
+
+  // Parse every statement in body.
+  // When encoutering a variable, look it up in the variable map.
+  for (const google::protobuf::Struct &primitive : bmv2_action.primitives()) {
+    if (primitive.fields().count("op") != 1 ||
+        primitive.fields().count("parameters") != 1) {
+      return absl::Status(
+          absl::StatusCode::kInvalidArgument,
+          absl::StrFormat("Primitive statement in action %s should contain 'op'"
+                          ", 'parameters', found %s",
+                          pdpi_action.preamble().name(),
+                          primitive.DebugString()));
+    }
+
+    const std::string &operation = primitive.fields().at("op").string_value();
+    // TODO(babman): Maybe bring back the enum and use switch-case here? discuss
+    Statement *statement = action_impl->add_action_body();
+    if (operation == "assign") {
+      AssignmentStatement *assignment = statement->mutable_assignment();
+      const google::protobuf::Value &params =
+          primitive.fields().at("parameters");
+      if (params.kind_case() != google::protobuf::Value::kListValue ||
+          params.list_value().values_size() != 2) {
+        return absl::Status(
+            absl::StatusCode::kInvalidArgument,
+            absl::StrFormat("Assignment statement in action %s must contain 2 "
+                            "parameters, found %s",
+                            pdpi_action.preamble().name(),
+                            primitive.DebugString()));
+      }
+
+      ASSIGN_OR_RETURN(
+          LValue lvalue,
+          ExtractLValue(params.list_value().values(0), variable_map));
+      ASSIGN_OR_RETURN(
+          RValue rvalue,
+          ExtractRValue(params.list_value().values(1), variable_map));
+      assignment->mutable_left()->CopyFrom(lvalue);
+      assignment->mutable_right()->CopyFrom(rvalue);
+    } else {
+      return absl::Status(
+          absl::StatusCode::kUnimplemented,
+          absl::StrFormat("Unsupported statement in action %s, found %s",
+                          pdpi_action.preamble().name(),
+                          primitive.DebugString()));
+    }
+
+    // Parse source_info struct into its own protobuf.
+    // Applies to all types of statements.
+    if (primitive.fields().count("source_info") != 1) {
+      return absl::Status(
+          absl::StatusCode::kInvalidArgument,
+          absl::StrFormat(
+              "Statement in action %s does not have source_info, found %s",
+              pdpi_action.preamble().name(), primitive.DebugString()));
+    }
+
+    ASSIGN_OR_RETURN(
+        bmv2::SourceLocation source_location,
```

Comment:
Just in case my comment was too concise to be understandable: what I was saying is that you can write directly to the statement, no need to create a temporary variable `source_location`.